### PR TITLE
Property assignment uses parent type annotation

### LINKED ISSUE #28492: Property assignment declarations on TS functions ignore declared type
**TypeScript Version:**  3.2.0-dev.20181110

**Search Terms:**
* Properties declarations on functions
* Function properties

**Code**

```ts
interface ExampleNum {
    (): boolean
    num: 123;
}

interface ExampleStr {
    (): boolean
    str: "ABC";
}

const exampleNum: ExampleNum = () => true;
exampleNum.num = 123;

const exampleStr: ExampleStr = () => true;
exampleStr.str = "ABC";
```

**Expected behavior:**
`exampleNum.num` to have type of `123`
`exampleStr.str` to have type of `"ABC"`
The function to remain as `(): boolean` instead of being coerced into `(): true`

**Actual behavior:**
> Type '{ (): true; num: number; }' is not assignable to type 'ExampleNum'.
>   Types of property 'num' are incompatible.
>     Type 'number' is not assignable to type '123'.

> Type '{ (): true; str: string; }' is not assignable to type 'ExampleStr'.
>   Types of property 'str' are incompatible.
>     Type 'string' is not assignable to type '"ABC"'.

**Playground Link:** 
http://www.typescriptlang.org/play/#src=interface%20ExampleNum%20%7B%0A%20%20%20%20()%3A%20boolean%0A%20%20%20%20num%3A%20123%3B%0A%7D%0A%0Ainterface%20ExampleStr%20%7B%0A%20%20%20%20()%3A%20boolean%0A%20%20%20%20str%3A%20%22ABC%22%3B%0A%7D%0A%0Aconst%20exampleNum%3A%20ExampleNum%20%3D%20()%20%3D%3E%20true%3B%0AexampleNum.num%20%3D%20123%3B%0A%0Aconst%20exampleStr%3A%20ExampleStr%20%3D%20()%20%3D%3E%20true%3B%0AexampleStr.str%20%3D%20%22ABC%22%3B

**Related Issues:** 
No related issues found relating to losing type safety

### PR DESCRIPTION
**Pull request by sandersn**
1. A property assignment declaration can get its type from a type annotation directly on the assignment, or from a type annotation on its parent. Since this is an assignment, you might assume that the parent's type would be available as a contextual type, but declarations are not contextually typed, even assignment declarations. This PR explicitly checks the parent's declaration for a type annotation. Fixes #28492
1. Property assignment declarations previously assumed only JSDoc. Now they use the standard function, getEffectiveTypeAnnotationNode, to look for type annotations. The name of the calling function is now getAnnotatedTypeForAssignmentDeclaration to reflect that.
2. getWidenedType**From**AssignmentDeclaration is now getWidenedType**For**AssignmentDeclaration. Finally!
3. I always search for `Whoa` instead of `getApparentTypeOfContextualType`, but it was spelled `Woah`.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -5396,7 +5396,7 @@ namespace ts {
             return undefined;
         }
 
-        function getWidenedTypeFromAssignmentDeclaration(symbol: Symbol, resolvedSymbol?: Symbol) {
+        function getWidenedTypeForAssignmentDeclaration(symbol: Symbol, resolvedSymbol?: Symbol) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: this is change (3)

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -5478,8 +5478,8 @@ namespace ts {
             return type;
         }
 
-        function getJSDocTypeFromAssignmentDeclaration(declaredType: Type | undefined, expression: Expression, _symbol: Symbol, declaration: Declaration) {
-            const typeNode = getJSDocType(expression.parent);
+        function getAnnotatedTypeForAssignmentDeclaration(declaredType: Type | undefined, expression: Expression, symbol: Symbol, declaration: Declaration) {
+            const typeNode = getEffectiveTypeAnnotationNode(expression.parent);
```
Comment (sandersn) [inline src/compiler/checker.ts]: this is change (2)

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -5489,6 +5489,13 @@ namespace ts {
                     errorNextVariableOrPropertyDeclarationMustHaveSameType(/*firstDeclaration*/ undefined, declaredType, declaration, type);
                 }
             }
```
Comment (sandersn) [inline src/compiler/checker.ts]: this is change (1)

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -19084,7 +19091,7 @@ namespace ts {
         }
 
         /**
```
Comment (sandersn) [inline src/compiler/checker.ts]: this is change (4)

**Comment by sandersn** [comment]
Google-fighting "whoa" vs "woah" gives 56M vs 38M. But google-fighting is a tool of hack linguists.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -5478,8 +5478,8 @@ namespace ts {
             return type;
         }
 
-        function getJSDocTypeFromAssignmentDeclaration(declaredType: Type | undefined, expression: Expression, _symbol: Symbol, declaration: Declaration) {
-            const typeNode = getJSDocType(expression.parent);
+        function getAnnotatedTypeForAssignmentDeclaration(declaredType: Type | undefined, expression: Expression, symbol: Symbol, declaration: Declaration) {
```
Comment (andrewbranch) [inline src/compiler/checker.ts]: What `expression` gets passed to this function? The binary expression with the `=` operator? Or one side of it? Your change makes sense, but I’m trying to visualize what was happening before without actually running the code.

Comment (sandersn) [reply src/compiler/checker.ts]: Yep. Here is the relevant code that finds the expression from a JS declaration:

```ts
                const expression = (isBinaryExpression(declaration) || isCallExpression(declaration)) ? declaration :
                    isPropertyAccessExpression(declaration) ? isBinaryExpression(declaration.parent) ? declaration.parent : declaration :
                        undefined;
```

That chained-nested ternary is scary. Maybe we should figure out a way to unnest it.

**Pull request merged**